pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -50,6 +59,166 @@
 )
 
 
+def _is_mcp_run_python_server(toolset: AbstractToolset[Any]) -> bool:
+    """Check if a toolset is an mcp-run-python server to avoid deadlock."""
+    command = getattr(toolset, 'command', None)
+    args = getattr(toolset, 'args', None)
+    if command is not None and args is not None:
+        command_str = ' '.join([str(command)] + [str(arg) for arg in args])
+        return 'mcp-run-python' in command_str
+    return False
+
+
+class _FilteredToolset(AbstractToolset[Any]):
+    """Toolset wrapper that excludes mcp-run-python servers to prevent deadlock."""
+
+    def __init__(self, wrapped: AbstractToolset[Any]):
+        from .toolsets.combined import CombinedToolset
+
+        self.wrapped = wrapped
+
+        # If it's a CombinedToolset, filter out mcp-run-python toolsets before any get_tools() calls
+        if isinstance(wrapped, CombinedToolset):
+            filtered_toolsets: list[AbstractToolset[Any]] = []
+            for toolset in wrapped.toolsets:
+                if not _is_mcp_run_python_server(toolset):
+                    filtered_toolsets.append(toolset)
+            # Create a new CombinedToolset with filtered toolsets
+            self.wrapped = CombinedToolset(toolsets=filtered_toolsets)
+        elif _is_mcp_run_python_server(wrapped):
+            # If the wrapped toolset itself is mcp-run-python, return empty tools
+            self._is_empty = True
+        else:
+            self._is_empty = False
+
+    async def get_tools(self, ctx: RunContext[Any]) -> dict[str, ToolsetTool[Any]]:
+        """Get tools excluding those from mcp-run-python servers."""
+        if hasattr(self, '_is_empty') and self._is_empty:
+            return {}
+
+        return await self.wrapped.get_tools(ctx)
+
+    async def call_tool(
+        self, name: str, tool_args: dict[str, Any], ctx: RunContext[Any], tool: ToolsetTool[Any]
+    ) -> Any:
+        """Call tool using the wrapped toolset."""
+        return await self.wrapped.call_tool(name, tool_args, ctx, tool)
+
+    def apply(self, visitor: Callable[[AbstractToolset[Any]], None]) -> None:
+        """Apply visitor to the wrapped toolset."""
+        self.wrapped.apply(visitor)
+
+
+async def _call_tool_for_elicitation(
+    toolset: AbstractToolset[Any],
+    tool_name: str,
+    tool_arguments: dict[str, Any],
+    ctx: RunContext[Any],
+) -> Any:
+    """Call a tool for elicitation using standard toolset API."""
+    available_tools = await toolset.get_tools(ctx)
+    if tool_name not in available_tools:
+        raise ValueError(f'Tool {tool_name} not found in toolset')
+
+    tool = available_tools[tool_name]
+    return await toolset.call_tool(tool_name, tool_arguments, ctx, tool)
```

Comment:
Note that this will not perform argument validation. I think we'll need to build a `ToolManager` and use `handle_call`